engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/AutoscopedConstructorTest.java

```diff
@@ -267,4 +267,44 @@ public void wrongConstructorNameYieldsTypeError() {
           e.getMessage().contains("Type_Error"));
     }
   }
+
+  @Test
+  public void simpleAnyCheck() {
+    var code =
+        """
+    import Standard.Base.Any.Any
+
+    type A
+        Typed x:Any
+
+    t = ..Typed ..My_Other
+    materialize v:A = v
+
+    create = materialize t
+    """;
+
+    var create = ctx.eval("enso", code).invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create");
+
+    assertEquals("A", create.getMetaObject().getMetaSimpleName());
+  }
+
+  @Test
+  public void simpleAnyOrACheck() {
+    var code =
+        """
+    import Standard.Base.Any.Any
+
+    type A
+        Typed (x:Any|A)
+
+    t = ..Typed ..My_Other
+    materialize v:A = v
+
+    create = materialize t
+    """;
+
+    var create = ctx.eval("enso", code).invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create");
+
+    assertEquals("A", create.getMetaObject().getMetaSimpleName());
+  }
```

Comment:
What happens if I pass `(..Typed 42)` to `foo` which has
```
foo (x : A|Any) = ...
```
?

Intuitively, while it _does have Any_ it also has `A` so it seems like maybe it would be good to try to materialize it?

Reply:
Could we do it as a rule that:
1. if there's no `Any`, always materialize
2. if it is only `Any` (or equivalent to it e.g. `Any|Any` etc.), never materialize,
3. if it is a sum type `A | B | Any` then try to materialize trying `A` or `B`, but if it fails, don't report a failure but instead 'fallback' to `Any` i.e. pass on the un-materialized type.

What do you think?

Reply:
As soon as there is `Any` in the signature, then no type check and no materialization happens.

Reply:
I know that this is the current behaviour, but I'm asking what do you think about this alternative one?

If we ever have a signature `Column_Selector | Any`, the current behaviour will force us to write 'hacks' like:
```
my_function (something : Column_Selector | Any) =
    resolved_something = Panic.catch No_Such_Conversion (something : Column_Selector) _->something
    ...
```
whereas this could work automatically.

But I guess we can also revisit that once we have use-cases like this, so I think for now this is indeed all we need.

Reply:
FYI: [Text | Any](https://github.com/enso-org/enso/blob/783276ca3403f54906cbc4658bcd212613a4e1b6/distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Errors.enso#L35).

Reply:
> FYI: [Text | Any](https://github.com/enso-org/enso/blob/783276ca3403f54906cbc4658bcd212613a4e1b6/distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Errors.enso#L35).

Yes I know we are using `X | Any` types in some places. That's why I was asking about this more general rule.

While the type `X | Any` is isomorphic to just `Any`, we use it to server a kind of documentation purpose. Often it is fields that can be one of several things: the example you linked is a good one showing this - we want to express that `cause` can either be a `Text` message explaining the error OR it can be another error/exception type that is the cause of this error (where the type `Any` expresses that, as we don't have an umbrella type for errors like `Throwable` or `Exception` in Java).

Not sure if in this particular place this is a _good API_, I imagine it may be written differently. But there are places where I think it makes sense - e.g. operations on `DB_Column` have signature `+ : DB_Column | Any -> DB_Column` - this tries to express that there is special semantics for `DB_Column` - if you do `column1 + column2` it zips the two columns and if you do `column1 + any_other_object` it applies `_ + any_other_object` to each element of `column1`. Of course this is better documented in the comments, but the type tries to express this kind of special behaviour and I think it's helpful.